src/map/script.cpp

```diff
@@ -24180,12 +24181,42 @@ BUILDIN_FUNC(mail){
 			}
 		}
 
+		// Refine
+		if (!script_hasdata(st, 9)) {
+			break;
+		}
+
+		data = script_getdata(st, 9);
+
+		if (!mail_sub(st, data, sd, 9, &name, &start, &end, &id)) {
+			return SCRIPT_CMD_FAILURE;
+		}
+
+		for (i = 0; i < num_items && start < end; i++, start++) {
+			struct item_data* item = itemdb_exists(msg.item[i].nameid);
+
+			ref = (int32)__64BPRTSIZE(get_val2(st, reference_uid(id, start), reference_getref(data)));
+
+			script_removetop(st, -1, 0);
+
+			if (item->type == IT_WEAPON || item->type == IT_ARMOR || item->type == IT_SHADOWGEAR) {
```

Comment:
You might want to add a check to make sure the item is able to be refined. IE:
```suggestion
			if (!item->flag.no_refine && (item->type == IT_WEAPON || item->type == IT_ARMOR || item->type == IT_SHADOWGEAR)) {
```